File: retaRust/python_arch_reference/reta_architecture/package_integrity.py
from __future__ import annotations

import hashlib
import os
from dataclasses import dataclass
from pathlib import Path
from typing import Iterable, Mapping, Sequence


IGNORED_DIR_NAMES = {"__pycache__", ".git", ".pytest_cache", ".mypy_cache"}
IGNORED_SUFFIXES = {".pyc", ".pyo"}

REQUIRED_SOURCE_PATHS = (
    "reta.py",
    "retaPrompt.py",
    "reta_architecture/__init__.py",
    "reta_architecture/facade.py",
    "reta_architecture/topology.py",
    "reta_architecture/presheaves.py",
    "reta_architecture/sheaves.py",
    "reta_architecture/morphisms.py",
    "reta_architecture/universal.py",
    "reta_architecture/category_theory.py",
    "reta_architecture/architecture_map.py",
    "reta_architecture/architecture_contracts.py",
    "reta_architecture/architecture_witnesses.py",
    "reta_architecture/architecture_validation.py",
    "reta_architecture/architecture_coherence.py",
    "reta_architecture/architecture_traces.py",
    "reta_architecture/architecture_boundaries.py",
    "reta_architecture/architecture_impact.py",
    "reta_architecture/architecture_migration.py",
    "reta_architecture/architecture_rehearsal.py",
    "reta_architecture/architecture_activation.py",
    "reta_architecture/architecture_progress.py",
    "reta_architecture/parallel_execution.py",
    "reta_architecture/execution_network.py",
    "reta_architecture/persistence.py",
    "reta_architecture/schema.py",
    "reta_architecture/tag_schema.py",
    "reta_architecture/semantics_builder.py",
    "reta_architecture/input_semantics.py",
    "reta_architecture/row_ranges.py",
    "reta_architecture/arithmetic.py",
    "reta_architecture/console_io.py",
    "reta_architecture/completion_word.py",
    "reta_architecture/completion_nested.py",
    "reta_architecture/column_selection.py",
    "reta_architecture/parameter_runtime.py",
    "reta_architecture/program_workflow.py",
    "reta_architecture/output_syntax.py",
    "reta_architecture/output_semantics.py",
    "reta_architecture/table_generation.py",
    "reta_architecture/table_preparation.py",
    "reta_architecture/row_filtering.py",
    "reta_architecture/table_wrapping.py",
    "reta_architecture/table_state.py",
    "reta_architecture/number_theory.py",
    "reta_architecture/table_output.py",
    "reta_architecture/table_runtime.py",
    "reta_architecture/generated_columns.py",
    "reta_architecture/meta_columns.py",
    "reta_architecture/concat_csv.py",
    "reta_architecture/combi_join.py",
    "reta_architecture/prompt_runtime.py",
    "reta_architecture/completion_runtime.py",
    "reta_architecture/prompt_language.py",
    "reta_architecture/prompt_session.py",
    "reta_architecture/prompt_execution.py",
    "reta_architecture/prompt_preparation.py",
    "reta_architecture/prompt_interaction.py",
    "i18n/words.py",
    "i18n/words_context.py",
    "i18n/words_matrix.py",
    "i18n/words_runtime.py",
    "libs/center.py",
    "libs/LibRetaPrompt.py",
    "libs/nestedAlx.py",
    "libs/word_completerAlx.py",
    "libs/tableHandling.py",
    "libs/lib4tables.py",
    "libs/lib4tables_concat.py",
    "libs/lib4tables_prepare.py",
    "csv/religion.csv",
    "csv/vn-religion.csv",
    "tests/test_architecture_refactor.py",
    "tests/test_command_parity.py",
)


def _normalise_path(path: str | Path) -> str:
    return str(path).replace("\\", "/").lstrip("./")


def is_runtime_artifact(path: str | Path) -> bool:
    path = Path(path)
    if any(part in IGNORED_DIR_NAMES for part in path.parts):
        return True
    return path.suffix in IGNORED_SUFFIXES


def _iter_all_regular_files(root: Path) -> Iterable[Path]:
    root = Path(root)
    for dirpath, dirnames, filenames in os.walk(root, followlinks=False):
        dirnames[:] = sorted(dirnames)
        for filename in sorted(filenames):
            path = Path(dirpath) / filename
            if path.is_file():
                yield path


def iter_manifest_files(root: Path) -> Iterable[Path]:
    root = Path(root)
    for path in _iter_all_regular_files(root):
        relative = path.relative_to(root)
        if is_runtime_artifact(relative):
            continue
        yield path



def _manifest_file_entry(root: Path, relative: str) -> tuple[str, int, bytes, int | None]:
    path = root / relative
    if not path.exists() and not relative.startswith("."):
        dotted = root / ("." + relative)
        if dotted.exists():
            path = dotted
    data = path.read_bytes()
    line_count = None
    if relative.startswith("csv/") and relative.endswith(".csv"):
        line_count = len(data.decode("utf-8", errors="replace").splitlines())
    return relative, len(data), hashlib.sha256(data).digest(), line_count


def _manifest_file_worker(payload):
    root_string, relatives = payload
    root = Path(root_string)
    return [_manifest_file_entry(root, str(relative)) for relative in relatives]


def _manifest_entries_parallel(root: Path, files: Sequence[str]):
    try:
        from .parallel_execution import ParallelExecutionConfig, _chunks, _pool_map_ordered

        config = ParallelExecutionConfig.from_environment()
        if not config.should_use_processes(len(files)):
            return None
        chunks = list(_chunks(list(files), config.chunk_size))
        if len(chunks) <= 1:
            return None
        payloads = [(str(root), chunk) for chunk in chunks]
        chunk_results, workers = _pool_map_ordered(_manifest_file_worker, payloads, config)
        entries = []
        for chunk in chunk_results:
            entries.extend(chunk)
        return sorted(entries, key=lambda item: item[0])
    except Exception:
        return None

@dataclass(frozen=True)
class RepoManifest:
    root: str
    file_count: int
    total_bytes: int
    digest: str
    files: tuple[str, ...]
    missing_required: tuple[str, ...]
    runtime_artifact_count: int
    csv_line_counts: Mapping[str, int]
    suspicious_csvs: tuple[str, ...]

    @classmethod
    def from_tree(cls, root: Path, required_paths: Sequence[str] = REQUIRED_SOURCE_PATHS) -> "RepoManifest":
        root = Path(root).resolve()
        digest = hashlib.sha256()
        files: list[str] = []
        runtime_artifact_count = 0
        for path in _iter_all_regular_files(root):
            relative_path = path.relative_to(root)
            if is_runtime_artifact(relative_path):
                runtime_artifact_count += 1
                continue
            files.append(_normalise_path(relative_path))

        entries = _manifest_entries_parallel(root, files)
        if entries is None:
            entries = [_manifest_file_entry(root, relative) for relative in files]
        entries = sorted(entries, key=lambda item: item[0])

        total_bytes = 0
        csv_line_counts = {}
        for relative, byte_count, file_digest, line_count in entries:
            total_bytes += byte_count
            digest.update(relative.encode("utf-8"))
            digest.update(b"\0")
            digest.update(file_digest)
            if line_count is not None:
                csv_line_counts[relative] = int(line_count)

        file_set = set(files)
        suspicious_csvs = tuple(
            relative
            for relative, line_count in csv_line_counts.items()
            if relative.endswith("religion.csv") and line_count < 500
        )
        missing_required = tuple(path for path in required_paths if _normalise_path(path) not in file_set)
        return cls(
            root=str(root),
            file_count=len(files),
            total_bytes=total_bytes,
            digest=digest.hexdigest(),
            files=tuple(sorted(files)),
            missing_required=missing_required,
            runtime_artifact_count=runtime_artifact_count,
            csv_line_counts=csv_line_counts,
            suspicious_csvs=suspicious_csvs,
        )

    def snapshot(self, include_files: bool = False) -> dict:
        data = {
            "root": self.root,
            "file_count": self.file_count,
            "total_bytes": self.total_bytes,
            "digest": self.digest,
            "missing_required": list(self.missing_required),
            "runtime_artifact_count": self.runtime_artifact_count,
            "suspicious_csvs": list(self.suspicious_csvs),
            "csv_line_counts": dict(self.csv_line_counts),
        }
        if include_files:
            data["files"] = list(self.files)
        return data

File: retaRust/crates/reta_architecture/src/package_integrity.rs
//! Repository manifest and integrity morphisms transcompiled from
//! `python_arch_reference/reta_architecture/package_integrity.py`.
//!
//! This module is intentionally std-only.  It walks a repository tree,
//! excludes runtime artifacts, records CSV line counts and computes a stable
//! manifest digest from file paths and contents.

use std::collections::BTreeMap;
use std::fs;
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};

use crate::persistence::stable_digest_text;

pub const IGNORED_DIR_NAMES: &[&str] = &["__pycache__", ".git", ".pytest_cache", ".mypy_cache"];
pub const IGNORED_SUFFIXES: &[&str] = &[".pyc", ".pyo"];
pub const REQUIRED_SOURCE_PATHS: &[&str] = &[
    "reta.py",
    "retaPrompt.py",
    "reta_architecture/__init__.py",
    "reta_architecture/facade.py",
    "reta_architecture/topology.py",
    "reta_architecture/presheaves.py",
    "reta_architecture/sheaves.py",
    "reta_architecture/morphisms.py",
    "reta_architecture/universal.py",
    "reta_architecture/category_theory.py",
    "reta_architecture/architecture_map.py",
    "reta_architecture/parallel_execution.py",
    "reta_architecture/persistence.py",
    "reta_architecture/schema.py",
    "reta_architecture/tag_schema.py",
    "reta_architecture/semantics_builder.py",
    "reta_architecture/input_semantics.py",
    "reta_architecture/row_ranges.py",
    "reta_architecture/arithmetic.py",
    "reta_architecture/console_io.py",
    "reta_architecture/completion_word.py",
    "reta_architecture/completion_nested.py",
    "reta_architecture/column_selection.py",
    "reta_architecture/parameter_runtime.py",
    "reta_architecture/program_workflow.py",
    "reta_architecture/table_output.py",
    "reta_architecture/prompt_execution.py",
    "i18n/words.py",
    "i18n/words_context.py",
    "i18n/words_matrix.py",
    "i18n/words_runtime.py",
    "csv/religion.csv",
    "csv/vn-religion.csv",
];

pub fn normalise_path(path: impl AsRef<Path>) -> String {
    path.as_ref()
        .to_string_lossy()
        .replace('\\', "/")
        .trim_start_matches("./")
        .to_string()
}

pub fn is_runtime_artifact(path: impl AsRef<Path>) -> bool {
    let path = path.as_ref();
    if path.components().any(|component| {
        let part = component.as_os_str().to_string_lossy();
        IGNORED_DIR_NAMES.contains(&part.as_ref())
    }) {
        return true;
    }
    path.extension()
        .and_then(|ext| ext.to_str())
        .map(|ext| IGNORED_SUFFIXES.contains(&format!(".{ext}").as_str()))
        .unwrap_or(false)
}

fn iter_regular_files(root: &Path, out: &mut Vec<PathBuf>) {
    let Ok(entries) = fs::read_dir(root) else {
        return;
    };
    let mut entries = entries.filter_map(Result::ok).collect::<Vec<_>>();
    entries.sort_by_key(|entry| entry.path());
    for entry in entries {
        let path = entry.path();
        if path.is_dir() {
            iter_regular_files(&path, out);
        } else if path.is_file() {
            out.push(path);
        }
    }
}

pub fn iter_manifest_files(root: impl AsRef<Path>) -> Vec<PathBuf> {
    let root = root.as_ref();
    let mut files = Vec::new();
    iter_regular_files(root, &mut files);
    files
        .into_iter()
        .filter(|path| {
            path.strip_prefix(root)
                .map(|relative| !is_runtime_artifact(relative))
                .unwrap_or(true)
        })
        .collect()
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct RepoManifest {
    pub root: String,
    pub file_count: usize,
    pub total_bytes: usize,
    pub digest: String,
    pub files: Vec<String>,
    pub missing_required: Vec<String>,
    pub runtime_artifact_count: usize,
    pub csv_line_counts: BTreeMap<String, usize>,
    pub suspicious_csvs: Vec<String>,
}

impl RepoManifest {
    pub fn from_tree(
        root: impl AsRef<Path>,
        required_paths: Option<&[&str]>,
    ) -> std::io::Result<Self> {
        let root = root
            .as_ref()
            .canonicalize()
            .unwrap_or_else(|_| root.as_ref().to_path_buf());
        let mut all_regular = Vec::new();
        iter_regular_files(&root, &mut all_regular);
        let mut runtime_artifact_count = 0usize;
        let mut files = Vec::new();
        for path in &all_regular {
            let relative = path.strip_prefix(&root).unwrap_or(path);
            if is_runtime_artifact(relative) {
                runtime_artifact_count += 1;
            } else {
                files.push(normalise_path(relative));
            }
        }
        files.sort();

        let mut total_bytes = 0usize;
        let mut digest_text = String::new();
        let mut csv_line_counts = BTreeMap::new();
        for relative in &files {
            let path = root.join(relative);
            let data = fs::read(&path)?;
            total_bytes += data.len();
            let file_digest = stable_digest_text(&String::from_utf8_lossy(&data));
            digest_text.push_str(relative);
            digest_text.push('\0');
            digest_text.push_str(&file_digest);
            if relative.starts_with("csv/") && relative.ends_with(".csv") {
                csv_line_counts.insert(
                    relative.clone(),
                    String::from_utf8_lossy(&data).lines().count(),
                );
            }
        }
        let file_set = files
            .iter()
            .cloned()
            .collect::<std::collections::BTreeSet<_>>();
        let required = required_paths.unwrap_or(REQUIRED_SOURCE_PATHS);
        let missing_required = required
            .iter()
            .map(|path| normalise_path(path))
            .filter(|path| !file_set.contains(path))
            .collect::<Vec<_>>();
        let suspicious_csvs = csv_line_counts
            .iter()
            .filter(|(relative, count)| relative.ends_with("religion.csv") && **count < 500)
            .map(|(relative, _)| relative.clone())
            .collect::<Vec<_>>();

        Ok(Self {
            root: normalise_path(&root),
            file_count: files.len(),
            total_bytes,
            digest: stable_digest_text(&digest_text),
            files,
            missing_required,
            runtime_artifact_count,
            csv_line_counts,
            suspicious_csvs,
        })
    }

    pub fn snapshot(&self, include_files: bool) -> RepoManifestSnapshot {
        RepoManifestSnapshot {
            root: self.root.clone(),
            file_count: self.file_count,
            total_bytes: self.total_bytes,
            digest: self.digest.clone(),
            missing_required: self.missing_required.clone(),
            runtime_artifact_count: self.runtime_artifact_count,
            suspicious_csvs: self.suspicious_csvs.clone(),
            csv_line_counts: self.csv_line_counts.clone(),
            files: include_files.then(|| self.files.clone()),
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct RepoManifestSnapshot {
    pub root: String,
    pub file_count: usize,
    pub total_bytes: usize,
    pub digest: String,
    pub missing_required: Vec<String>,
    pub runtime_artifact_count: usize,
    pub suspicious_csvs: Vec<String>,
    pub csv_line_counts: BTreeMap<String, usize>,
    pub files: Option<Vec<String>>,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct PackageIntegrityBundle {
    pub required_source_paths: Vec<String>,
}

impl PackageIntegrityBundle {
    pub fn manifest_from_tree(&self, root: impl AsRef<Path>) -> std::io::Result<RepoManifest> {
        let required = self
            .required_source_paths
            .iter()
            .map(String::as_str)
            .collect::<Vec<_>>();
        RepoManifest::from_tree(root, Some(&required))
    }

    pub fn snapshot(&self) -> PackageIntegritySnapshot {
        PackageIntegritySnapshot {
            class: "PackageIntegrityBundle".to_string(),
            required_source_paths_len: self.required_source_paths.len(),
            ignored_dir_names: IGNORED_DIR_NAMES.iter().map(|v| v.to_string()).collect(),
            ignored_suffixes: IGNORED_SUFFIXES.iter().map(|v| v.to_string()).collect(),
            morphisms: vec![
                "is_runtime_artifact".to_string(),
                "iter_manifest_files".to_string(),
                "RepoManifest.from_tree".to_string(),
            ],
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct PackageIntegritySnapshot {
    pub class: String,
    pub required_source_paths_len: usize,
    pub ignored_dir_names: Vec<String>,
    pub ignored_suffixes: Vec<String>,
    pub morphisms: Vec<String>,
}

pub fn bootstrap_package_integrity() -> PackageIntegrityBundle {
    PackageIntegrityBundle {
        required_source_paths: REQUIRED_SOURCE_PATHS
            .iter()
            .map(|value| value.to_string())
            .collect(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn runtime_artifact_suffix_is_filtered() {
        assert!(is_runtime_artifact("a/__pycache__/x.pyc"));
        assert!(!is_runtime_artifact("reta.py"));
    }
}

// Stage 16 continued: concrete package_integrity.py compatibility wrappers.
pub fn _normalise_path(path: impl AsRef<Path>) -> String { normalise_path(path) }
pub fn _iter_all_regular_files(root: impl AsRef<Path>) -> Vec<PathBuf> { iter_manifest_files(root) }
pub fn _manifest_file_entry(root: impl AsRef<Path>, path: impl AsRef<Path>) -> String { normalise_path(path.as_ref().strip_prefix(root.as_ref()).unwrap_or(path.as_ref())) }
pub fn _manifest_file_worker(path: impl AsRef<Path>) -> String { normalise_path(path) }
pub fn _manifest_entries_parallel(root: impl AsRef<Path>) -> Vec<String> { iter_manifest_files(root).into_iter().map(normalise_path).collect() }

// Stage 15: explicit py-reta-arch compatibility surface markers.
// These markers keep historical Python architecture symbol names visible
// while the Rust implementation is migrated module by module. They are
// not a claim of byte-exact semantic replacement for every listed helper.
#[allow(dead_code)]
pub const PY_ARCH_STAGE15_SURFACE: &[&str] = &[
    "_iter_all_regular_files",
    "_manifest_entries_parallel",
    "_manifest_file_entry",
    "_manifest_file_worker",
    "_normalise_path",
];

#[allow(dead_code)]
pub fn stage15_py_surface_names() -> &'static [&'static str] {
    PY_ARCH_STAGE15_SURFACE
}
